core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java

```diff
@@ -4175,98 +4179,491 @@ private RelNode convertUpdate(SqlUpdate call) {
+      Integer totalOffset, RelOptTable destTable) {
+
+
+    // Convert the identifiers in each of the update's target column list into a map of
+    // strings -> idx. This map helps us to convert the orderings during the later iteration.
+    // for information on why we need this, see extractUpdateExprs
+
+    HashMap<SqlInsert, HashMap<String, Integer>> insertToTargetColumnSet = new HashMap<>();
+    for (int insertCallIdx = 0; insertCallIdx < insertCallList.size(); insertCallIdx++) {
+      SqlInsert curInsertCall = (SqlInsert) insertCallList.get(insertCallIdx);
+      HashMap<String, Integer> curMap = new HashMap<String, Integer>();
+
+      @Nullable SqlNodeList curInsertTargetColumnList = curInsertCall.getTargetColumnList();
 
-      level1InsertExprs =
-          ((LogicalProject) insertRel.getInput(0)).getProjects();
-      if (insertRel.getInput(0).getInput(0) instanceof LogicalProject) {
-        level2InsertExprs =
-            ((LogicalProject) insertRel.getInput(0).getInput(0))
-                .getProjects();
+      if (curInsertTargetColumnList == null) {
+        //If curInsertTargetColumnList returns NULL, then it means that we're inserting
+        // every column in the destination table
+        for (int curColExprIdx = 0;
+             curColExprIdx < destTable.getRowType().getFieldCount(); curColExprIdx++) {
+          RelDataTypeField field = destTable.getRowType().getFieldList().get(curColExprIdx);
+          curMap.put(field.getName(), curColExprIdx);
+        }
+      } else {
+        for (int curColExprIdx = 0;
+             curColExprIdx < curInsertTargetColumnList.size(); curColExprIdx++) {
+          SqlIdentifier id = (SqlIdentifier) curInsertTargetColumnList.get(curColExprIdx);
+          RelDataTypeField field =
+              SqlValidatorUtil.getTargetField(
+                  destTable.getRowType(), typeFactory, id, catalogReader, destTable);
+          assert field != null : "column " + id.toString() + " not found";
+          curMap.put(field.getName(), curColExprIdx);
+        }
       }
-      nLevel1Exprs = level1InsertExprs.size();
+
+      insertToTargetColumnSet.put(curInsertCall, curMap);
+    }
+
+    // The list of conditions. This is returned from this function in order to filter no ops.
+    List<RexNode> caseConditions = new ArrayList<>();
+
+    // The list of column values to put into the table
+    // The outer list should have length of the number of data columns
+    // the inner list should have length equal to the number of conditions
+    // IE, if condition X were the first expression to evaluate to TRUE, the expression that
+    // column Y should evaluate to is caseValues[Y][X]
+    List<List<RexNode>> caseValues = new ArrayList<>();
+
+    List<String> destTableFieldNames = destTable.getRowType().getFieldNames();
+
+    for (int destColIdx = 0; destColIdx < destTableFieldNames.size(); destColIdx++) {
+      caseValues.add(new ArrayList<>());
     }
```

Comment:
Correct me if I'm wrong, but doesn't this mean that size of `caseValues` ends up being (`destTableFieldNames.size()` + 1) since you initialized `caseValues` with an ArrayList already?